### PR TITLE
Wip optimization

### PR DESCRIPTION
**Pull request by justinpombrio**

### CODE REVIEW

File: src/reducers.js [hunk 1]
```diff
@@ -17,7 +17,7 @@ export const reducer = (
     case 'SET_FOCUS':
       return {...state, focusId: action.focusId};
     case 'SET_AST':
-      return {...state, ast: action.ast};
+      return {...state, ast: action.ast, collapsedList: state.collapsedList.filter(action.ast.getNodeById)};
```
[the remark this answers is not in the document]
Comment (justinpombrio) [reply src/reducers.js]: ignoring

File: src/nodes.jsx [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import * as P from './pretty';
-import React from 'react';
+import React, {Component} from 'react';
```
[the remark this answers is not in the document]
Comment (justinpombrio) [reply src/nodes.jsx]: already fixed

File: src/components/BlockComponent.js [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
-  //   //   console.log('abort on ', oldValue);
-  //   // }
-  //   return shouldUpdate;
-  // }
+export default class BlockComponent extends Component {
+  shouldComponentUpdate(props, state) {
+    // NOTE: don't care about the node since the patching algorithm will deal
+    // with the update already
+    const {node: newValue, ...newProps} = props;
+    const {node: oldValue, ...oldProps} = this.props;
```
[the remark this answers is not in the document]
Comment (justinpombrio) [reply src/components/BlockComponent.js]: eh. It's specified in the subclasses of Blockcomponent, no big deal here.

File: src/ast.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import React from 'react';
+import React, {Component} from 'react';
```
[the remark this answers is not in the document]
Comment (justinpombrio) [reply src/ast.js]: already fixed.

File: src/ui/BlockEditor.js [hunk 5]
```diff
@@ -417,22 +416,13 @@ class BlockEditor extends Component {
   }
 
   renderPortals = () => {
-    const portals = [];
+    let portals;
```
[the remark this answers is not in the document]
Comment (schanzer) [reply src/ui/BlockEditor.js]: The latest commit uses `startOperation()` and `endOperation()`, so this should be resolved.

File: src/ui/BlockEditor.js [hunk 6]
```diff
@@ -417,22 +416,13 @@ class BlockEditor extends Component {
...
-
-        // NOTE(Oak): we need to clear all markers up front to prevent
-        // overlapping the marker issue
-        marker.clear();
-      }
-      for (const r of this.props.ast.rootNodes) {
-        portals.push(<ToplevelBlock key={r.id} node={r} />);
-      }
-      if (this.props.hasQuarantine) portals.push(<ToplevelBlockEditable key="-1" />);
-      setTimeout(() => { SHARED.cm.refresh(); console.log('refreshed CM'); }, 100);
```
[the remark this answers is not in the document]
Comment (schanzer) [reply src/ui/BlockEditor.js]: Added it back in the latest commit

**Pull request merged**